Reland 263531@main without Speedometer regression

nt1m opened the pull request:
#### 0fcccfdd39fb158061c2eb8eb6dc1d0c220f568d
<pre>
Reland 263531@main without Speedometer regression
<a href="https://bugs.webkit.org/show_bug.cgi?id=256674">https://bugs.webkit.org/show_bug.cgi?id=256674</a>
rdar://109234330

Reviewed by Simon Fraser.

The ElementRuleCollector.cpp change was causing the perf regression, reland without it for now, since it&apos;s not strictly necessary to fix the repaint issue.

Also fix similar repaint issues with background/outline which were not fixed by the original patch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-parent-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-parent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-parent-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-parent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint.html: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresRepaint const):
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
(WebCore::StyleBackgroundData::isEquivalentForPainting const):

Canonical link: <a href="https://commits.webkit.org/264218@main">https://commits.webkit.org/264218@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/383dea690a26a2cf7c7aba172fb2c91c99c27a19

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7003 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7244 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7424 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/8616 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7237 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7012 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/8506 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7178 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10146 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/7799 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/6463 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/8719 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/5167 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/6371 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14144 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/6836 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/6465 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/9312 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6939 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/5686 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6307 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/6281 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1669 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10501 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1281,19 +1281,19 @@ bool RenderStyle::changeRequiresRepaint(const RenderStyle& other, OptionSet<Styl
+        if (m_nonInheritedData->surroundData.ptr() != other.m_nonInheritedData->surroundData.ptr())
+            return m_nonInheritedData->surroundData->border != other.m_nonInheritedData->surroundData->border;
+
+        if (m_nonInheritedData->surroundData->border.hasVisibleCurrentColor())
```
You also need to check `other.m_nonInheritedData->surroundData->border`, and if you do that isn't this equivalent to the old code?

nt1m replied:
ah right I wrote this code too fast, this should be done under the assumption that `m_nonInheritedData->surroundData->border == other.m_nonInheritedData->surroundData->border;`

smfr commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -1281,19 +1281,19 @@ bool RenderStyle::changeRequiresRepaint(const RenderStyle& other, OptionSet<Styl
-            if (!m_nonInheritedData->backgroundData->isEquivalentForPainting(*other.m_nonInheritedData->backgroundData, currentColorDiffers))
-                return true;
-        }
+    bool nonInheritedDataDiffers = m_nonInheritedData.ptr() != other.m_nonInheritedData.ptr();
```
`nonInheritedDataDiffers` is not an accurate name; even if the pointers are different, the data might be the same. We have `arePointingToEqualData()` to help with this comparison.

smfr commented on Source/WebCore/rendering/style/StyleBackgroundData.h:
```diff
@@ -42,7 +42,8 @@ class StyleBackgroundData : public RefCounted<StyleBackgroundData> {
 
-    bool isEquivalentForPainting(const StyleBackgroundData&, bool currentColorDiffers) const;
+    bool hasVisibleCurrentColor() const;
+    bool isEquivalentForPainting(const StyleBackgroundData*, bool currentColorDiffers) const;
```
Not sure why you change the arg to a pointer. Also, I'd prefer all the StyleFoo classes use the same pattern.

nt1m replied:
This is so I can move the pointer comparison in the method

nt1m replied:
Ah I can use arePointingToEqualData() if I have a ref

nt1m committed:
Reland 263531@main without Speedometer regression
https://bugs.webkit.org/show_bug.cgi?id=256674
rdar://109234330

Reviewed by Simon Fraser.

The ElementRuleCollector.cpp change was causing the perf regression, reland without it for now, since it's not strictly necessary to fix the repaint issue.

Also fix similar repaint issues with background/outline which were not fixed by the original patch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-parent-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint-parent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-background-repaint.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-parent-expected.xht: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint-parent.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-backgrounds/color-mix-currentcolor-outline-repaint.html: Added.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::changeRequiresRepaint const):
* Source/WebCore/rendering/style/StyleBackgroundData.cpp:
(WebCore::StyleBackgroundData::isEquivalentForPainting const):

Canonical link: https://commits.webkit.org/264218@main
